File: examples/contrib/assignment.py
# Copyright 2010 Hakan Kjellerstrand hakank@gmail.com
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Assignment problem in Google CP Solver.

Winston 'Operations Research', Assignment Problems, page 393f
(generalized version with added test column)

Compare with the following models:
* Comet   : http://www.hakank.org/comet/assignment.co
* ECLiPSE : http://www.hakank.org/eclipse/assignment.ecl
* Gecode  : http://www.hakank.org/gecode/assignment.cpp
* MiniZinc: http://www.hakank.org/minizinc/assignment.mzn
* Tailor/Essence': http://www.hakank.org/tailor/assignment.eprime
* SICStus: http://hakank.org/sicstus/assignment.pl

This model was created by Hakan Kjellerstrand (hakank@gmail.com)
Also see my other Google CP Solver models:
http://www.hakank.org/google_or_tools/
"""

from ortools.constraint_solver.python import constraint_solver as cp


def main(cost, rows, cols):
    # Create the solver.
    solver = cp.Solver("n-queens")

    #
    # data
    #

    # declare variables
    total_cost = solver.new_int_var(0, 100, "total_cost")
    x = []
    for i in range(rows):
        t = []
        for j in range(cols):
            t.append(solver.new_int_var(0, 1, "x[%i,%i]" % (i, j)))
        x.append(t)
    x_flat = [x[i][j] for i in range(rows) for j in range(cols)]

    #
    # constraints
    #

    # total_cost
    solver.add(
        total_cost
        == solver.sum(
            [solver.weighted_sum(x_row, cost_row) for (x_row, cost_row) in zip(x, cost)]
        )
    )

    # exacly one assignment per row, all rows must be assigned
    [
        solver.add(solver.sum([x[row][j] for j in range(cols)]) == 1)
        for row in range(rows)
    ]

    # zero or one assignments per column
    [
        solver.add(solver.sum([x[i][col] for i in range(rows)]) <= 1)
        for col in range(cols)
    ]

    objective = solver.minimize(total_cost, 1)

    #
    # solution and search
    #
    solution = solver.assignment()
    solution.add(x_flat)
    solution.add(total_cost)

    # db: DecisionBuilder
    db = solver.phase(
        x_flat,
        cp.IntVarStrategy.INT_VAR_SIMPLE,
        cp.IntValueStrategy.ASSIGN_MIN_VALUE,
    )

    solver.new_search(db, [objective])
    num_solutions = 0
    while solver.next_solution():
        print("total_cost:", total_cost.value())
        for i in range(rows):
            for j in range(cols):
                print(x[i][j].value(), end=" ")
            print()
        print()

        for i in range(rows):
            print("Task:", i, end=" ")
            for j in range(cols):
                if x[i][j].value() == 1:
                    print(" is done by ", j)
        print()

        num_solutions += 1
    solver.end_search()

    print()
    print("num_solutions:", num_solutions)
    print("failures:", solver.num_failures)
    print("branches:", solver.num_branches)
    print("WallTime:", solver.wall_time_ms)


# Problem instance
# hakank: I added the fifth column to make it more
#         interesting
rows = 4
cols = 5
cost = [[14, 5, 8, 7, 15], [2, 12, 6, 5, 3], [7, 8, 3, 9, 7], [2, 4, 6, 10, 1]]

if __name__ == "__main__":
    main(cost, rows, cols)

File: examples/contrib/assignment.cs
//
// Copyright 2012 Hakan Kjellerstrand
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

using System;
using System.Collections;
using System.IO;
using System.Linq;
using System.Text.RegularExpressions;
using Google.OrTools.ConstraintSolver;

public class Assignment
{
    /**
     *
     * Assignment problem
     *
     * From Wayne Winston "Operations Research",
     * Assignment Problems, page 393f
     * (generalized version with added test column)
     *
     * See  See http://www.hakank.org/or-tools/assignment.py
     *
     */
    private static void Solve()
    {
        Solver solver = new Solver("Assignment");

        //
        // data
        //

        // Problem instance
        // hakank: I added the fifth column to make it more
        //         interesting
        int rows = 4;
        int cols = 5;
        int[,] cost = { { 14, 5, 8, 7, 15 }, { 2, 12, 6, 5, 3 }, { 7, 8, 3, 9, 7 }, { 2, 4, 6, 10, 1 } };

        //
        // Decision variables
        //
        IntVar[,] x = solver.MakeBoolVarMatrix(rows, cols, "x");
        IntVar[] x_flat = x.Flatten();

        //
        // Constraints
        //

        // Exacly one assignment per row (task),
        // i.e. all rows must be assigned with one worker
        for (int i = 0; i < rows; i++)
        {
            solver.Add((from j in Enumerable.Range(0, cols) select x[i, j]).ToArray().Sum() == 1);
        }

        // At most one assignments per column (worker)
        for (int j = 0; j < cols; j++)
        {
            solver.Add((from i in Enumerable.Range(0, rows) select x[i, j]).ToArray().Sum() <= 1);
        }

        // Total cost
        IntVar total_cost =
            (from i in Enumerable.Range(0, rows) from j in Enumerable.Range(0, cols) select(cost[i, j] * x[i, j]))
                .ToArray()
                .Sum()
                .Var();

        //
        // objective
        //
        OptimizeVar objective = total_cost.Minimize(1);

        //
        // Search
        //
        DecisionBuilder db = solver.MakePhase(x_flat, Solver.INT_VAR_DEFAULT, Solver.INT_VALUE_DEFAULT);

        solver.NewSearch(db, objective);

        while (solver.NextSolution())
        {
            Console.WriteLine("total_cost: {0}", total_cost.Value());
            for (int i = 0; i < rows; i++)
            {
                for (int j = 0; j < cols; j++)
                {
                    Console.Write(x[i, j].Value() + " ");
                }
                Console.WriteLine();
            }
            Console.WriteLine();
            Console.WriteLine("Assignments:");
            for (int i = 0; i < rows; i++)
            {
                Console.Write("Task " + i);
                for (int j = 0; j < cols; j++)
                {
                    if (x[i, j].Value() == 1)
                    {
                        Console.WriteLine(" is done by " + j);
                    }
                }
            }
            Console.WriteLine();
        }

        Console.WriteLine("\nSolutions: {0}", solver.Solutions());
        Console.WriteLine("WallTime: {0}ms", solver.WallTime());
        Console.WriteLine("Failures: {0}", solver.Failures());
        Console.WriteLine("Branches: {0} ", solver.Branches());

        solver.EndSearch();
    }

    public static void Main(String[] args)
    {
        Solve();
    }
}
